### PR TITLE
Tests: create new fixtures to make writing booted/restored simpler

### PR DESCRIPTION
**Pull request by pb8o**
## Changes

This introduces a few new pytest fixtures to deal with CPU template differences, A/B tests, and booted/restored VMs.

In addition there's some cleanups in the affected tests.

## Reason

Making those tests simple to write is important so that we don't forget one case. In addition they become easier to write.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] I have read and understand [CONTRIBUTING.md][3].
- [x] I have run `tools/devtool checkstyle` to verify that the PR passes the
  automated style checks.
- [x] I have described what is done in these changes, why they are needed, and
  how they are solving the problem in a clear and encompassing way.
- [x] I have updated any relevant documentation (both in code and in the docs)
  in the PR.
- [x] I have mentioned all user-facing changes in `CHANGELOG.md`.
- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] When making API changes, I have followed the
  [Runbook for Firecracker API changes][2].
- [x] I have tested all new and changed functionalities in unit tests and/or
  integration tests.
- [x] I have linked an issue to every new `TODO`.

______________________________________________________________________

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: tests/integration_tests/functional/test_cpu_features_aarch64.py
Comment (zulinx86) [inline tests/integration_tests/functional/test_cpu_features_aarch64.py]: `parse_cpu_flags()` not defined anywhere but will be removed in a subsequent commit. Is it used for temporary purpose?

File: tests/integration_tests/functional/test_cpu_features_host_vs_guest.py
Comment (zulinx86) [inline tests/integration_tests/functional/test_cpu_features_host_vs_guest.py:85]: nit: The variable names of the expected feat sets are different between cases. How about unifying the variable name (like `host_minus_guest_5_10`, `host_only_5_10`, or something)?

File: tests/integration_tests/functional/test_rng.py
Comment (zulinx86) [inline tests/integration_tests/functional/test_rng.py:220]: This doesn't test work on restored uVMs? Or we just don't need to test it on restored ones?

File: tests/conftest.py
Comment (zulinx86) [reply tests/conftest.py:495]: ah, does that mean "function"? I'd prefer a longer name (`func`) or another name (like `constructor`). wdyt?

**Review by zulinx86 — COMMENTED** [review]
I really love this series of refactoring! I only left some minor comments.

File: tests/framework/microvm.py
Comment (pb8o) [reply tests/framework/microvm.py:940]: Now I don't quite remember, but there are some tests that read `kernel_file` and were not working when using the "restored" fixture. Now I don't know which test exactly was it. This was at the beginning but it has since changed substantially. I am not 100% sure it is needed anymore.

File: tests/conftest.py
Comment (pb8o) [reply tests/conftest.py:495]: Yeah I was thinking `uvm_ctor` to abbreviate it a bit :D

File: tests/framework/microvm.py
Comment (pb8o) [reply tests/framework/microvm.py:778]: Done!

File: tests/integration_tests/functional/test_cpu_features_aarch64.py
Comment (pb8o) [reply tests/integration_tests/functional/test_cpu_features_aarch64.py]: Hmm I think there was a rebase gone wrong here.

File: tests/framework/microvm.py
Comment (pb8o) [reply tests/framework/microvm.py]: Done!

**Review by zulinx86 — APPROVED** [review]
Looks good to me to merge :)

Do you think it's worth updating the description for pytest fixtures in the test README file?
https://github.com/firecracker-microvm/firecracker/blob/81bae9f8284abc2ab6778746bb60563d224c8025/tests/README.md?plain=1#L300-L308

**Comment by zulinx86** [comment]
It'd be nice to run it in non-PR mode before merging.

File: tests/integration_tests/security/test_vulnerabilities.py
Comment (zulinx86) [inline tests/integration_tests/security/test_vulnerabilities.py:250]: If the template is a static one, `cpu_template` is None, so should pass `cpu_template_name` here.
```suggestion
            uvm_any.cpu_template_name
```

File: tests/integration_tests/security/test_vulnerabilities.py
Comment (zulinx86) [inline tests/integration_tests/security/test_vulnerabilities.py:89]: `cpu_template` is None if no template or static templates. Should use `cpu_template_name` here instead.
```suggestion
    def expected_vulnerabilities(self, cpu_template_name):
        """
        There is a REPTAR exception reported on INTEL_ICELAKE when spectre-meltdown-checker.sh
        script is run inside the guest from below the tests:
            test_spectre_meltdown_checker_on_guest and
            test_spectre_meltdown_checker_on_restored_guest
        The same script when run on host doesn't report the
        exception which means the instances are actually not vulnerable to REPTAR.
        The only reason why the script cannot determine if the guest
        is vulnerable or not because Firecracker does not expose the microcode
        version to the guest.
        The check in spectre_meltdown_checker is here:
            https://github.com/speed47/spectre-meltdown-checker/blob/0f2edb1a71733c1074550166c5e53abcfaa4d6ca/spectre-meltdown-checker.sh#L6635-L6637
        Since we have a test on host and the exception in guest is not valid,
        we add a check to ignore this exception.
        """
        if global_props.cpu_codename == "INTEL_ICELAKE" and cpu_template_name == "None":
```

**Review by zulinx86 — CHANGES_REQUESTED** [review]

File: tests/integration_tests/security/test_vulnerabilities.py
Comment (pb8o) [reply tests/integration_tests/security/test_vulnerabilities.py:233]: Yes, `guest_kernel` for example will be the same for `uvm_any` and `uvm_any_a`, because pytest caches the fixtures for a test. That's an important property. I will add it to the docstring.

https://docs.pytest.org/en/stable/how-to/fixtures.html#fixtures-can-be-requested-more-than-once-per-test-return-values-are-cached

File: tests/integration_tests/security/test_vulnerabilities.py
Comment (pb8o) [reply tests/integration_tests/security/test_vulnerabilities.py]: Good catch! Fixed

File: tests/framework/microvm.py
Comment (zulinx86) [reply tests/framework/microvm.py:940]: I misinterpreted how the metadata works. I also used it in #4942 and now understand it! One drawback from the maintainability point of view is the difficulty to find whether the metadata is really needed or not, since, when it used, it is referenced through `MicroVM` instance. But it doesn't have to be addressed now and is not a blocker to merge this PR :)

File: tests/integration_tests/functional/test_rng.py
Comment (zulinx86) [reply tests/integration_tests/functional/test_rng.py]: got it, leave it as it is here 👍

**Review by zulinx86 — APPROVED** [review]
LGTM! Thanks a lot for this series of changes again!

**Comment by pb8o** [comment]
> It'd be nice to run it in non-PR mode before merging.

I think at least the latest I have run it in both modes.

File: tests/integration_tests/security/test_vulnerabilities.py
Comment (Manciukic) [inline tests/integration_tests/security/test_vulnerabilities.py:68]: nit: should we assert this in the guest as well?

File: tests/integration_tests/security/test_nv.py
Comment (Manciukic) [inline tests/integration_tests/security/test_nv.py:20]: out of scope: the name of this test is so cryptic, I had to read the code to understand what was happening lol

File: tests/integration_tests/functional/test_cpu_features_aarch64.py
Comment (Manciukic) [inline tests/integration_tests/functional/test_cpu_features_aarch64.py:47]: nit: should we `strip()` the `stdout` as well to avoid trailing blank characters resulting in empty strings? or does `split()` already remove those?

File: tests/integration_tests/functional/test_cpu_features_aarch64.py
Comment (Manciukic) [inline tests/integration_tests/functional/test_cpu_features_aarch64.py]: ditto as above

File: tests/integration_tests/functional/test_cpu_features_host_vs_guest.py
Comment (Manciukic) [inline tests/integration_tests/functional/test_cpu_features_host_vs_guest.py:85]: nit: we could move this list out as well as we're adding Genoa soon

File: tests/integration_tests/functional/test_rng.py
Comment (Manciukic) [inline tests/integration_tests/functional/test_rng.py:12]: nit: update docstring? is this still a fixture?

File: tests/conftest.py
Comment (Manciukic) [inline tests/conftest.py:461]: out of curiosity, how do we override this?

**Review by Manciukic — DISMISSED** [review]

File: tests/integration_tests/functional/test_cpu_features_aarch64.py
Comment (pb8o) [reply tests/integration_tests/functional/test_cpu_features_aarch64.py:47]: Yeah, `.split()` behaves different than `.split(" ")` and does what you want commonly:

```
"     a   b    c    ".split()
['a', 'b', 'c']
```

File: tests/conftest.py
Comment (pb8o) [reply tests/conftest.py:461]: For example:

```
@pytest.mark.parametrize("vcpu_count", [1, 2, 7], indirect=True)
def test_new_test(uvm_any):
    pass
```

File: tests/integration_tests/security/test_vulnerabilities.py
Comment (pb8o) [reply tests/integration_tests/security/test_vulnerabilities.py:68]: Actually `check_output` should already be checking that, so it's not needed.

File: tests/integration_tests/security/test_nv.py
Comment (pb8o) [reply tests/integration_tests/security/test_nv.py:20]: Sorry that was my fault. I wouldn't be against giving it a more proper name.

File: tests/integration_tests/functional/test_cpu_features_host_vs_guest.py
Comment (pb8o) [reply tests/integration_tests/functional/test_cpu_features_host_vs_guest.py:85]: I agree, but will defer to that onboarding.

File: tests/integration_tests/security/test_vulnerabilities.py
Comment (pb8o) [reply tests/integration_tests/security/test_vulnerabilities.py]: Fixed

File: tests/framework/properties.py
Comment (pb8o) [reply tests/framework/properties.py:81]: Done. Yeah I actually thought this would be more involved but it's a small change.

File: tests/framework/properties.py
Comment (pb8o) [reply tests/framework/properties.py:82]: Done

File: tests/integration_tests/functional/test_cpu_features_aarch64.py
Comment (pb8o) [reply tests/integration_tests/functional/test_cpu_features_aarch64.py]: This changes later to `assert guest_feats == host_feats`, so will keep it as-is

File: tests/integration_tests/security/test_nv.py
Comment (pb8o) [reply tests/integration_tests/security/test_nv.py]: Done

File: tests/integration_tests/functional/test_cpu_features_aarch64.py
Comment (roypat) [reply tests/integration_tests/functional/test_cpu_features_aarch64.py]: Ah, missed that, works for me!

**Review by roypat — APPROVED** [review]

**Review by Manciukic — APPROVED** [review]

**Pull request merged**
